### PR TITLE
test: add discovery tests

### PR DESCRIPTION
**Pull request by baszalmstra**
Adds test for detecting how to configure a build backend from a source directory. Note that this code path is not _yet_ used in pixi itself. 

Test cases are stored as directories in the `tests/data/discovery` directory. The test code locates any files called `TEST_CASE` in that directory and performs a snapshot test in the parent directory of that file. This allows very easy addition of tests by simple creating a new folder with that file somewhere. I've added a large number of different edge cases that test workspace inheritance and proper error reporting. 

I also added additional tests for checking what happens if the source directory doesnt exist or when one depends directly on a `recipe.yaml` file.

### CODE REVIEW

**Review by Hofer-Julian — COMMENTED** [review]
Only had minimal comments.

Great work Bas 😺

File: tests/data/discovery/nested/TEST-CASE [hunk 1]
```diff
@@ -0,0 +1 @@
+Marks this directory as a test-case for discovery tests.
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply tests/data/discovery/nested/TEST-CASE:1]: Yes, I also want the case were its just a workspace (and error on it).

[the remark this answers is not in the document]
Comment (Hofer-Julian) [reply tests/data/discovery/nested/TEST-CASE:1]: For me, this intention would be a bit more clear if you'd just add a separate test case for it :)

[the remark this answers is not in the document]
Comment (baszalmstra) [reply tests/data/discovery/nested/TEST-CASE:1]: will do

**Review by Hofer-Julian — APPROVED** [review]
Let's bring it iiiiiin 🛠️

File: crates/pixi_build_frontend/src/discovery.rs [hunk 2]
```diff
@@ -270,15 +275,18 @@ impl DiscoveredBackend {
...
-        source_dir: &Path,
+        source_dir: PathBuf,
         channel_config: &ChannelConfig,
     ) -> Result<Option<Self>, DiscoveryError> {
-        for (&recipe_dir, &recipe_file) in VALID_RECIPE_DIRS.iter().zip(VALID_RECIPE_NAMES.iter()) {
-            let recipe_path = source_dir.join(recipe_dir).join(recipe_file);
-            if recipe_path.is_file() {
+        for (&recipe_dir, &recipe_file) in VALID_RECIPE_DIRS
+            .iter()
+            .cartesian_product(VALID_RECIPE_NAMES.iter())
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply crates/pixi_build_frontend/src/discovery.rs:286]: Thanks to the tests 😄
